### PR TITLE
Download export: angular => knockout

### PR DESCRIPTION
**Pull request by orangejenny**
This combines https://github.com/dimagi/commcare-hq/pull/22197 and https://github.com/dimagi/commcare-hq/pull/22158 into one PR, because they're going through QA together and keeping them separate is getting difficult to manage.

@esoergel @nickpell You've both reviewed the early commits. The first unreviewed commit is [
Added support for list of exports](https://github.com/dimagi/commcare-hq/commit/2fee0de2375cbea83d755cca1a751a7f6fd2b29d)  

[QA ticket](http://manage.dimagi.com/default.asp?285073)

### CODE REVIEW

**Comment by orangejenny** [comment]
@esoergel @nickpell Responded to comments from https://github.com/dimagi/commcare-hq/pull/22158, ready for re-review.

File: corehq/apps/export/static/export/js/download_export.js [hunk 1]
```diff
@@ -1,7 +1,15 @@
+/**
+ *  UI for the page to download an export (form, case, or SMS; single or bulk).
+ *
+ *  Contains two models:
+ *      downloadFormModel: Controls the form where the user specifies a date range, etc. Also contains
+ *          functionality to download the multimedia associated with an export (form exports only).
+ *      downloadProgressModel: Controls the progress bar, etc. once the user has clicked 'Prepare Export'.
+ *          Includes functionality to email the user when the export is done, rather than them waiting for it.
+ */
```
Comment (esoergel) [inline corehq/apps/export/static/export/js/download_export.js]: this is very helpful, thanks!

File: corehq/apps/export/static/export/js/download_export.js [hunk 2]
```diff
@@ -160,19 +156,19 @@ hqDefine('export/js/download_export', function () {
...
         };
 
-        self._handlePrepareError = function (data) {
+        self.handleError = function (data) {
             if (data && data.error) {
                 // The server returned an error message.
                 self.prepareExportError(data.error);
             } else {
-                self.prepareExportError("default");
+                self.prepareExportError(gettext("Sorry, there was a problem reaching the server. Please try again."));
```
Comment (esoergel) [inline corehq/apps/export/static/export/js/download_export.js]: This is a bit off-topic, but do you know if we have any sort of convention for reusing these sorts of messages?  I imagine every UI has some subtle variation of the same few messages, each of which needs to be translated independently.

File: corehq/apps/export/views/download.py [hunk 3]
```diff
@@ -289,72 +288,72 @@ def prepare_custom_export(request, domain):
...
-    except ExportFormValidationException:
-        return json_response({
-            'error': _("Form did not validate."),
-        })
     except XlsLengthException:
         return json_response({
             'error': _('This file has more than 256 columns, which is not supported by xls. '
                        'Please change the output type to csv or xlsx to export this file.')
         })
+
```
Comment (esoergel) [inline corehq/apps/export/views/download.py]: Good job on fixing up the exception handling in the last few commits

File: corehq/apps/export/templates/export/download_export.html [hunk 4]
```diff
@@ -57,7 +57,7 @@
     {% registerurl 'prepare_custom_export' domain %}
     {% registerurl 'prepare_form_multimedia' domain %}
 
-    <div id="download-export-form">
+    <div id="download-export-form" class="ko-template">
```
Comment (esoergel) [inline corehq/apps/export/templates/export/download_export.html]: How's that work?  Does it stay hidden till initialized or something?

**Review by esoergel — APPROVED** [review]
Looks great!

File: corehq/apps/export/static/export/js/download_export.js (hunk 2 above)
Comment (orangejenny) [reply corehq/apps/export/static/export/js/download_export.js]: I don't think we do, but it's a good idea and probably not difficult to implement: I'd imagine a module in `hqwebapp` just full of string constants that we'd use as error messages.

File: corehq/apps/export/templates/export/download_export.html (hunk 4 above)
Comment (orangejenny) [reply corehq/apps/export/templates/export/download_export.html]: yep, exactly: https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/hqwebapp/static/hqwebapp/js/hq.helpers.js#L157

**Pull request merged**